hydradx/notebooks/Stableswap/ImpermanentLossAnalysis4Assets.ipynb

```diff
@@ -0,0 +1,573 @@
+    "            },\n",
+    "            amplification=10,\n",
+    "            trade_fee=0\n",
+    "        )\n",
+    "    },\n",
+    "    agents = {\n",
+    "        'LP': Agent(\n",
+    "            holdings={tkn: 1 for tkn in assets},\n",
+    "            trade_strategy=invest_all('stableswap')\n",
+    "        ),\n",
+    "        'Trader': Agent(\n",
+    "            holdings={tkn: 10000000000000 for tkn in assets}\n",
+    "        )\n",
+    "    },\n",
+    "    external_market={tkn: assets[tkn]['usd price'] for tkn in assets}\n",
+    ")\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2023-07-06T20:00:31.857764207Z",
+     "start_time": "2023-07-06T20:00:31.681435289Z"
+    },
+    "tags": []
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "0.000995444390312422 247276.40443434505\n",
+      "156.87953617143862 -99990.00099990002\n",
+      "0.000977326252482079 247647.57879376572\n",
+      "199.71967208980743 -96686.31486433979\n",
+      "0.0009877770884735417 247944.91939021496\n",
+      "199.94530000067437 -92802.53170641157\n",
+      "0.000980612059546324 248217.79521216897\n",
+      "199.67437907817265 -90032.92489873343\n",
+      "0.000995282647861672 248441.28060711964\n",
+      "199.67207475942692 -88073.92649009275\n",
+      "0.0009588261640137211 248664.96721939923\n",
+      "199.42914906713446 -86684.69418227047\n"
+     ]
+    }
+   ],
+   "source": [
+    "invested_state = initial_state.agents['LP'].trade_strategy.execute(\n",
+    "    initial_state.copy(),\n",
+    "    agent_id='LP'\n",
+    ")\n",
+    "amps = [25, 50, 100, 200, 400, 800]\n",
+    "trade_range = [[200000, -100000]] * len(amps)\n",
+    "# print(trade_range)\n",
+    "for i, amp in enumerate(amps):\n",
+    "    for j, target_price in enumerate([0.001, 200]):\n",
+    "        invested_state.pools['stableswap'].amplification = amp\n",
+    "        try_state = invested_state.copy()\n",
+    "        spot_price = [1, 2000][j]\n",
+    "        while (\n",
+    "            spot_price > target_price \n",
+    "            and not try_state.pools['stableswap'].fail\n",
+    "        ):\n",
+    "            # find the correct trade size to get the price we want\n",
+    "            if j == 0:\n",
+    "                trade_range[i][j] *= 1.0001\n",
+    "            elif j == 1:\n",
+    "                trade_range[i][j] /= 1.0001\n",
+    "            try_state = invested_state.copy()\n",
+    "            try_state.pools['stableswap'].swap_one(\n",
+    "                agent=try_state.agents['Trader'],\n",
+    "                tkn_sell='USDA',\n",
+    "                quantity=trade_range[i][j]\n",
+    "            )\n",
+    "            spot_price = try_state.pools['stableswap'].spot_price('USDA', numeraire='USDD')\n",
```

Comment:
this call fails for me, causing the notebook to fail? spot_price doesn't appear to have a "numeraire" input

Reply:
fixed, I just hadn't pushed it yet